### PR TITLE
Adds support for filtering PropertyHelpers by property visibility (new

### PR DESCRIPTION
**Pull request by rynowak**
keyword).

We have a version of this already in route value dictionary that can be
merged, we're going to use this in model binding/validation.

### CODE REVIEW

**Comment by rynowak** [comment]
@halter73

File: src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs [hunk 1]
```diff
@@ -23,9 +24,14 @@ internal class PropertyHelper
...
 
-        private static readonly ConcurrentDictionary<Type, PropertyHelper[]> ReflectionCache =
+        // Using an array rather than IEnumerable, as target will be called on the hot path numerous times.
+        private static readonly ConcurrentDictionary<Type, PropertyHelper[]> PropertiesCache =
             new ConcurrentDictionary<Type, PropertyHelper[]>();
 
+        private static readonly ConcurrentDictionary<Type, PropertyHelper[]> VisiblePropertiesCache =
+            new ConcurrentDictionary<Type, PropertyHelper[]>();
+
+
```
Comment (harshgMSFT) [inline src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs]: extra line.

File: src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs [hunk 2]
```diff
@@ -69,7 +75,45 @@ public static PropertyHelper[] GetProperties(object instance)
         /// </returns>
         public static PropertyHelper[] GetProperties(Type type)
         {
-            return GetProperties(type, CreateInstance, ReflectionCache);
+            return GetProperties(type, CreateInstance, PropertiesCache);
+        }
+
+        /// <summary>
+        /// <para>
+        /// Creates a caches fast property helpers that expose getters for every non-hidden get property
```
Comment (harshgMSFT) [inline src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs]: ?huh?  is the description correct

File: src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs [hunk 3]
```diff
@@ -192,6 +236,77 @@ private static PropertyHelper CreateInstance(PropertyInfo property)
...
+                }
+
+                // If this property was declared on a base type then look for the definition closest to the
+                // the type to see if we should include it.
+                var ignoreProperty = false;
+
+                // Walk up the hierarchy until we find the type that actally declares this
+                // PropertyInfo.
+                var currentType = type.GetTypeInfo();
+                while (currentType != null && currentType != propertyHelper.Property.DeclaringType.GetTypeInfo())
```
Comment (harshgMSFT) [inline src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs]: suggest adding a variable for `propertyHelper.Property.DeclaringType`

**Comment by harshgMSFT** [comment]
:watch:  for the description ..

File: src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs]: should be "creates and caches" this is basically copied from the other overloads

Comment (harshgMSFT) [inline src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs]: yup ... .

**Comment by harshgMSFT** [comment]
:shipit:  after fixing the description.

**Pull request merged**